### PR TITLE
[browser] Add history erasing within periods. JB#55578

### PR DESCRIPTION
**Pull request by Maves1**
Add browsing history erasing within periods (last day, last week, last 4 weeks, everything).
"Last 4 weeks" is used here because the initial suggestion to use "Last month" is vague: we can set the period to {31, 30, 29, 28} days.

### CODE REVIEW

File: apps/browser/qml/pages/PrivacySettingsPage.qml [hunk 1]
```diff
@@ -118,6 +189,13 @@ Page {
         }
     }
 
+    ConfigurationValue {
+        id: clearHistoryPeriodConfig
+
+        key: "/apps/sailfish-browser/actions/clear_history_period"
```
Comment (rainemak) [inline apps/browser/qml/pages/PrivacySettingsPage.qml:195]: Same as https://github.com/sailfishos/sailfish-browser/pull/912#discussion_r717437831 CC @atatarov

Comment (rainemak) [reply apps/browser/qml/pages/PrivacySettingsPage.qml:195]: PR https://github.com/sailfishos/sailfish-browser/pull/912 merged, please rebase and update this accordingly.

File: apps/storage/dbworker.cpp [hunk 2]
```diff
@@ -489,6 +490,47 @@ void DBWorker::clearHistory()
...
+    for (iconIterator = historyIconsToRemove.begin(); iconIterator != historyIconsToRemove.end(); iconIterator++) {
+        FaviconManager::instance()->remove(QStringLiteral("history"), *iconIterator);
+    }
+
+    uint boundaryTimeT = QDateTime::currentDateTimeUtc().addDays(-days).toTime_t();
+
+    QSqlQuery query = prepare("DELETE FROM browser_history WHERE date > ?");
+    query.bindValue(0, boundaryTimeT);
+    execute(query);
+    removeAllTabs();
```
Comment (atatarov) [inline apps/storage/dbworker.cpp:529]: This call is duplicated
https://github.com/sailfishos/sailfish-browser/pull/914/files#diff-6ee2a3c1bdc5d546b02a363a605dfbc2fbe3287b0adf62f0208b49cc9a29f5c6R184

Comment (atatarov) [reply apps/storage/dbworker.cpp:529]: if (clearHistory.checked) {
                                                         Settings.clearHistory()
                                                         Settings.removeAllTabs()
                                                     }

File: apps/browser/qml/pages/PrivacySettingsPage.qml [hunk 3]
```diff
@@ -50,6 +51,49 @@ Page {
...
 
+                ComboBox {
+                     id: historyErasingComboBox
+
+                     width: parent.width
+                     //% "Clear browser history for"
+                     label: qsTrId("settings_browser-la-clear_history_period")
+                     value: {
+                         switch (currentIndex) {
+                         case 0:
```
Comment (atatarov) [inline apps/browser/qml/pages/PrivacySettingsPage.qml:62]: could you put it to something like this
`
var historyPeriods {
 0 : HistoryModel.Everything,
 1 : HistoryModel.Day,
 ....
}
`

File: apps/storage/dbworker.cpp [hunk 4]
```diff
@@ -477,9 +478,49 @@ void DBWorker::addHistoryEntry(const QString &url, const QString &title)
     }
 }
 
-void DBWorker::clearHistory()
+void DBWorker::clearHistory(int period)
 {
-    QSqlQuery query = prepare("DELETE FROM browser_history;");
+    QSqlQuery query;
+
+    if (period == 0) { // Delete everything
```
Comment (atatarov) [inline apps/storage/dbworker.cpp:485]: It could be cast to enum

File: apps/browser/qml/pages/PrivacySettingsPage.qml [hunk 5]
```diff
@@ -110,8 +154,20 @@ Page {
                         remorse = Remorse.popupAction(page, qsTrId("settings_browser-la-cleared_private_data"),
                                                  function() {
                                                      if (clearHistory.checked) {
-                                                         Settings.clearHistory()
-                                                         Settings.removeAllTabs()
+                                                         switch(historyPeriodIndex) {
+                                                         case 0:
+                                                             Settings.clearHistory(HistoryModel.Day)
```
Comment (atatarov) [inline apps/browser/qml/pages/PrivacySettingsPage.qml:159]: It could be like this
Settings.clearHistory(historyPeriods[historyPeriodIndex])

**Review by atatarov — APPROVED** [review]

File: apps/core/settingmanager.cpp [hunk 6]
```diff
@@ -12,6 +12,7 @@
 #include "dbmanager.h"
 #include "opensearchconfigs.h"
 #include "faviconmanager.h"
+#include "declarativehistorymodel.h"
```
Comment (llewelld) [inline apps/core/settingmanager.cpp:15]: It looks like this header include can be dropped now.

File: apps/storage/dbworker.h [hunk 7]
```diff
@@ -47,12 +47,13 @@ public slots:
     void goForward(int tabId);
     void goBack(int tabId);
     void getHistory(const QString &filter);
+    QList<Link> getHistoryQList();
```
Comment (llewelld) [inline apps/storage/dbworker.h:50]: Does this need to be a `public slot`? Could it even be made `private`?

**Comment by llewelld** [comment]
Thanks for making the QML changes @Maves1. This looks nice to me now; I gave it some light testing and it also worked as expected. I've just left a few very minor comments, but otherwise this looks good to me.

File: apps/storage/dbworker.cpp [hunk 8]
```diff
@@ -477,9 +478,49 @@ void DBWorker::addHistoryEntry(const QString &url, const QString &title)
...
+                historyIconsToRemove.remove(sanitizedUrl);
+            }
+        }
+
+        QSet<QString>::iterator iconIterator;
+        for (iconIterator = historyIconsToRemove.begin(); iconIterator != historyIconsToRemove.end(); iconIterator++) {
+            FaviconManager::instance()->remove(QStringLiteral("history"), *iconIterator);
+        }
+
+        uint boundaryTimeT = QDateTime::currentDateTimeUtc().addDays(-period).toTime_t();
```
[the remark this answers is not in the document]
Comment (Maves1) [reply apps/storage/dbworker.cpp:516]: We are saving history in UTC, so I think everything has to be fine here

**Review by llewelld — APPROVED** [review]
Thanks for making those small changes @Maves1. This is a really nice addition and LGTM.

**Comment by llewelld** [comment]
I don't think it should affect merging this PR, but it would be good to add some unit tests for this new functionality, for example to the [tst_declarativehistorymodel tests](https://github.com/sailfishos/sailfish-browser/tree/master/tests/auto/tst_declarativehistorymodel). They would probably  be pretty straightforward to add, but would also reassure me about those UTC edge-cases. Anyway, perhaps something to consider for a separate task.
